### PR TITLE
Implement simple JSON serializer

### PR DESCRIPTION
**Pull request by jdisanti**
This adds a JSON serializer for #161. At this point, I've vetted that it meets the requirements of the code generator, although the generator is not finished yet. Posting this as its own PR to keep the reviews bite sized.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rust-runtime/smithy-json/src/serialize.rs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        });
+        self
+    }
+
+    /// Writes a string to the array.
+    pub fn string(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        append_string(&mut self.json, value);
+        self
+    }
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:118]: I _might_ consider a small optimization here for serializing base64 data (which I know can be a hot path).

In the case of base64, we know ahead of time that we don't need to perform any escaping so in that case we could do something like `string_unchecked` that skipped the escape phase.

Comment (rcoh) [reply rust-runtime/smithy-json/src/serialize.rs:118]: (here and other `string` methods)

File: rust-runtime/smithy-json/src/serialize.rs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+fn append_string(json: &mut String, value: &str) {
+    json.push('"');
+    json.push_str(&escape_string(value));
+    json.push('"');
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:159]: can you add test coverage for `null`/`boolean`? All the other methods are covered

File: rust-runtime/smithy-json/src/serialize.rs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+    /// Writes a string `value` with the given `key`.
+    pub fn string(&mut self, key: &str, value: &str) -> &mut Self {
+        self.key(key);
+        append_string(&mut self.json, value);
+        self
+    }
+
+    /// Writes a number `value` (already converted to string) with the given `key`.
+    pub fn number(&mut self, key: &str, value: &str) -> &mut Self {
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:47]: I will leave this up to you, but I think we will probably want this library to handle numeric JSON conversion: https://github.com/awslabs/smithy-rs/blob/42c1f215a5f88cd58871117dee3a67fa5b0b318f/rust-runtime/smithy-types/src/lib.rs#L57

Otherwise, we'll need to handle it upstream in a bunch of different places

File: rust-runtime/smithy-json/src/escape.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    let mut escaped = String::new();
+    let (mut last, end) = (0, value.len());
+    for (index, chr) in value
+        .char_indices()
+        .filter(|(_index, chr)| ESCAPES.contains(chr))
+    {
+        escaped.push_str(&value[last..index]);
+        escaped.push_str(match chr {
+            '"' => "\\\"",
+            '\\' => "\\\\",
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/escape.rs:25]: nit: no test for this case

File: rust-runtime/smithy-json/src/escape.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            '\u{08}' => "\\b",
+            '\u{0C}' => "\\f",
+            '\n' => "\\n",
+            '\r' => "\\r",
+            '\t' => "\\t",
+            _ => unreachable!(),
+        });
+        last = index + 1;
+    }
+    escaped.push_str(&value[last..end]);
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/escape.rs:35]: can you add a test where there is data after the escape? As it is currently, the tests will pass if this line is deleted

File: rust-runtime/smithy-json/src/escape.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    #[test]
+    fn escape() {
+        assert_eq!("", escape_string("").as_ref());
+        assert_eq!("foo", escape_string("foo").as_ref());
+        assert_eq!("foo\\r\\n", escape_string("foo\r\n").as_ref());
+        assert_eq!(
+            "\\bf\\fo\\to\\r\\n",
+            escape_string("\u{08}f\u{0C}o\to\r\n").as_ref()
+        );
+        assert_eq!("\\\"test\\\"", escape_string("\"test\"").as_ref());
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/escape.rs:52]: nit: all these tests may be slightly more readable as raw string literals

Comment (rcoh) [reply rust-runtime/smithy-json/src/escape.rs:52]: (with the exception of the newline ones)

File: rust-runtime/smithy-json/src/escape.rs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    use proptest::proptest;
+    proptest! {
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/escape.rs:56]: could consider using `serde_json` as a dev dependency to validate that these escapes are valid JSON / unescape back (or waiting for your implementation of unescape is also fine)

**Review by rcoh — APPROVED** [review]
LGTM!

Code looks good, just noted a few places where test coverage could be improved.

File: rust-runtime/smithy-json/src/serialize.rs (hunk 3 above)
Comment (jdisanti) [reply rust-runtime/smithy-json/src/serialize.rs:47]: Oh, I wasn't aware of the `Number` type in `smithy-types`. That simplifies things considerably. Originally, I was thinking I would need to pull in the `num` crate in order to write a generic number function, or alternatively make a separate number function for each number type and have the code generator choose, but I settled for just calling `to_string()` everywhere this `number` function is called.

Will refactor.

File: rust-runtime/smithy-json/src/serialize.rs (hunk 1 above)
Comment (jdisanti) [reply rust-runtime/smithy-json/src/serialize.rs:118]: I like this idea.

File: rust-runtime/smithy-json/src/escape.rs (hunk 7 above)
Comment (jdisanti) [reply rust-runtime/smithy-json/src/escape.rs:56]: I like this idea. Added.

File: rust-runtime/smithy-json/src/serialize.rs (hunk 3 above)
Comment (jdisanti) [reply rust-runtime/smithy-json/src/serialize.rs:47]: Also moved the timestamp handling code out of the code gen and into the serializers so that it's not repeated everywhere.

File: rust-runtime/smithy-json/src/serialize.rs [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+    pub fn string(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        append_string(&mut self.json, value);
+        self
+    }
+
+    /// Writes a string `value` to the array without escaping it.
+    pub fn trusted_string(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        append_string(&mut self.json, value);
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:139]: should be `append_trusted_string`
```suggestion
        append_trusted_string(&mut self.json, value);
```

File: rust-runtime/smithy-json/src/serialize.rs [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+    json.push('"');
+    json.push_str(value);
+    json.push('"');
+}
+
+fn append_instant(json: &mut String, value: &Instant, format: Format) {
+    let formatted = value.fmt(format);
+    match format {
+        Format::EpochSeconds => json.push_str(&formatted),
+        _ => append_string(json, &formatted),
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:196]: this can be `append_trusted_string`

File: rust-runtime/smithy-json/src/serialize.rs [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+    match value {
+        Number::PosInt(value) => {
+            json.push_str(itoa::Buffer::new().format(value));
+        }
+        Number::NegInt(value) => {
+            json.push_str(itoa::Buffer::new().format(value));
+        }
+        Number::Float(value) => {
+            // If the value is NaN, Infinity, or -Infinity
+            if value.to_bits() & F64_EXPONENT_MASK == F64_EXPONENT_MASK {
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:212]: can you use `value.is_nan()`?

File: rust-runtime/smithy-json/src/serialize.rs [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+                json.push_str("null");
+            } else {
+                let mut buffer = ryu::Buffer::new();
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:215]: I would add a comment that `ryu::Buffer::new()` and `itoa::Buffer:new()` are cheap. I assumed this allocated until I read the code, might as well save some future person from trying to optimize this.

In fact, `itoa::write` just calls `itoa::Buffer::new()`!

File: rust-runtime/smithy-json/src/serialize.rs [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+
+    /// Writes a number `value` to the array.
+    pub fn number(&mut self, value: Number) -> &mut Self {
+        self.comma_delimit();
+        append_number(&mut self.json, value);
+        self
+    }
+
+    /// Writes an Instant `value` with the given `key` and `format`.
+    pub fn instant(&mut self, instant: &Instant, format: Format) -> &mut Self {
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:151]: cool idea to add this here! maybe we should do the same for smithy-xml

File: rust-runtime/smithy-json/src/serialize.rs [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+    }
+
+    /// Writes a number `value` to the array.
+    pub fn number(&mut self, value: Number) -> &mut Self {
+        self.comma_delimit();
+        append_number(&mut self.json, value);
+        self
+    }
+
+    /// Writes an Instant `value` with the given `key` and `format`.
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:150]:
```suggestion
    /// Writes an Instant `value` to this array with `format`.
```

File: rust-runtime/smithy-json/src/serialize.rs (hunk 8 above)
Comment (rcoh) [reply rust-runtime/smithy-json/src/serialize.rs:139]: I would probably renamed this `string_unchecked` — it's not that we trust/don't trust the string as much that we aren't going to check that it doesn't need to be escaped

File: rust-runtime/smithy-json/src/escape.rs [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        #[test]
+        fn matches_serde_json(s: String) {
+            let mut formatted = String::new();
+            formatted.push_str(r#"{"value":""#);
+            formatted.push_str(&escape_string(&s));
+            formatted.push_str("\"}");
+
+            assert_eq!(
+                serde_json::to_string(&TestStruct { value: s }).unwrap(),
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/escape.rs:74]: you don't need to wrap the value in a structure here, you can call `to_string` on a string directly: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dcc0f9d691e95c859419cc46cc2cc4e1

this applies here & can also simplify the float tests as well

File: rust-runtime/smithy-json/src/serialize.rs [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+use crate::escape::escape_string;
+use smithy_types::instant::Format;
+use smithy_types::{Instant, Number};
+
+pub struct JsonObjectWriter<'a> {
+    json: &'a mut String,
+    started: bool,
+}
+
+impl<'a> JsonObjectWriter<'a> {
```
Comment (rcoh) [inline rust-runtime/smithy-json/src/serialize.rs:15]: not sure how this currently plays out in codegen but I _might_ consider:

```rust
pub struct JsonValueWriter<'a> {
  json: &'a mut String,
}

// all writing of values goes through this one place
impl JsonValueWriter {
   pub fn null(self) {
     self.push_str("null");
   }
   pub fn boolean(&self, value) {
      // ...
   }
}

// then instead of having lots of duplicated methods:
impl JsonObjectWriter<'a> {
  pub fn key(&mut self, key: &str) -> JsonValueWriter {
    // write "key": with the comma tracking etc.
    JsonValueWriter { json: self.json }
  }
}

// then the usage would look like:
let mut obj_1 = JsonObjectWriter::new(&mut output);
obj_1.key("foo").boolean(true);
let mut nested = obj_1.key("nested").object();
nested.key("asdf").string("hello!");
nested.finish();
obj_1.finish();
```

File: rust-runtime/smithy-json/src/serialize.rs [hunk 16]
```diff
@@ -0,0 +1,451 @@
...
+        self.started = true;
+    }
+}
+
+fn append_string(json: &mut String, value: &str) {
+    append_trusted_string(json, &escape_string(value));
+}
+
+fn append_trusted_string(json: &mut String, value: &str) {
+    json.push('"');
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/smithy-json/src/serialize.rs:187]: I really like this. Will add.

[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/smithy-json/src/serialize.rs:187]: Slight problem though: it won't actually be `Cow::Borrowed` in this case:
```
object.string("line one\nline two");
```
Since that string will feed into `escape_str` to become `"line one\\nline two"` before making its way into `append_string_unchecked`, which then asserts escaping it is a `Cow::Borrowed`, except that that won't be the case, because running escaping on it again would change it to `"line one\\\\nline two"` to preserve the two slashes.

If we really want to be safe about it, I could create an enum that tracks the "escapedness" of the string, if you think that's worth it.

File: rust-runtime/smithy-json/src/serialize.rs (hunk 15 above)
Comment (jdisanti) [reply rust-runtime/smithy-json/src/serialize.rs:15]: I really like this idea. I'll simplify down to this in a follow up PR.

File: rust-runtime/smithy-json/src/serialize.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (rcoh) [reply rust-runtime/smithy-json/src/serialize.rs:187]: ah I see! I don't think that's worth it

[the remark this answers is not in the document]
Comment (rcoh) [reply rust-runtime/smithy-json/src/serialize.rs:187]: Or actually we can just not have `append_string` go through trusted string

[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/smithy-json/src/serialize.rs:187]: Yeah, that seems like a good compromise. Have some duplicated code to insert quotes for the added safety.
